Upgrade Chalk

jackh726 opened the pull request:
Things done in this PR:
- Upgrade Chalk to `0.11.0`
- Added compare-mode=chalk
- Bump rustc-hash in `librustc_data_structures` to `1.1.0` to match Chalk
- Removed `RustDefId` since the builtin type support is there
- Add a few more `FIXME(chalk)`s for problem spots I hit when running all tests with chalk
- Added some more implementation code for some newer builtin Chalk types (e.g. `FnDef`, `Array`)
- Lower `RegionOutlives` and `ObjectSafe` predicates
- Lower `Dyn` without the region
- Handle `Int`/`Float` `CanonicalVarKind`s
- Uncomment some Chalk tests that actually work now
- Remove the revisions in `src/test/ui/coherence/coherence-subtyping.rs` since they aren't doing anything different

r? @nikomatsakis

nikomatsakis commented on src/librustc_traits/chalk/db.rs:
```diff
@@ -175,9 +175,39 @@ impl<'tcx> chalk_solve::RustIrDatabase<RustInterner<'tcx>> for RustIrDatabase<'t
+            .filter_map(|wc| LowerInto::<Option<chalk_ir::QuantifiedWhereClause<RustInterner<'tcx>>>>::lower_into(wc, &self.interner)).collect();
+
+        let sig = self.tcx.fn_sig(def_id);
+        // FIXME(chalk): Why does this have a Binder
```
Because of late-bound regions -- I overlooked that, we should fix chalk to have a binder as well. For a function like

```rust
fn foo<'a>(x: &'a u32) { }
```

the `'a` is considered "late-bound" and doesn't show up in the generics of `foo`, but rather in its signature.

nikomatsakis replied:
We should probably not invoke `skip_binder` here but rather something like `no_bound_vars().expect("FIXME: late-bound fn parameters not supported in chalk")`

jackh726 replied:
That's weird. So do lifetimes never show up in the generics of functions?

jackh726 replied:
Ugh so any functions that take `&self` (e.g. in `inherent_impl`) are broken

nikomatsakis commented on src/librustc_traits/chalk/mod.rs:
```diff
@@ -199,7 +199,7 @@ crate fn evaluate_goal<'tcx>(
             Solution::Unique(_subst) => {
                 // FIXME(chalk): handle constraints
-                assert!(_subst.value.constraints.is_empty());
+                // assert!(_subst.value.constraints.is_empty());
```
maybe just delete this?

nikomatsakis commented on src/librustc_middle/traits/chalk.rs:
```diff
@@ -86,8 +73,8 @@ impl<'tcx> chalk_ir::interner::Interner for RustInterner<'tcx> {
-    type DefId = RustDefId;
-    type InternedAdtId = RustDefId;
+    type DefId = DefId;
+    type InternedAdtId = DefId;
```
Can we make this [`AdtDef<'tcx>`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/struct.AdtDef.html) instead?

jackh726 replied:
Can't do this now since `InternedAdtDef` has a `Copy` bound.

nikomatsakis commented on src/librustc_traits/chalk/db.rs:
```diff
@@ -125,62 +119,54 @@ impl<'tcx> chalk_solve::RustIrDatabase<RustInterner<'tcx>> for RustIrDatabase<'t
-            RustDefId::Adt(adt_def_id) => {
-                let adt_def = self.tcx.adt_def(adt_def_id);
+        let adt_def_id = struct_id.0;
+        let adt_def = self.tcx.adt_def(adt_def_id);
```
...then we don't need this call...

nikomatsakis commented on src/librustc_traits/chalk/db.rs:
```diff
@@ -381,25 +345,14 @@ impl<'tcx> chalk_solve::RustIrDatabase<RustInterner<'tcx>> for RustIrDatabase<'t
-                                    }
-                                }
+                    chalk_ir::TypeName::Adt(chalk_ir::AdtId(adt_def_id)) => {
+                        let adt_def = self.tcx.adt_def(adt_def_id);
```
...or this one, etc

bors commented:
:umbrella: The latest upstream changes (presumably #70107) made this pull request unmergeable. Please resolve the merge conflicts.

jackh726 commented:
@nikomatsakis addressed your changes
- Removed the commented out assert
- Changed `skip_binders` to `no_bound_vars().expect()`. This requires collecting into an intermediate `Vec`, since `inputs()` returns a slice, which doesn't impl `TypeFoldable`. But since this is going to be fixed on the Chalk side anyways, I don't think this is a problem right now. It did catch the fact that we were ignoring this in the `inherent_impl` test, so I commented out the bits of code that fail now :(
- Can't change `InternedAdtDef` to `AdtDef` since it has a `Copy` bound and `AdtDef` isn't.

nikomatsakis commented:
@bors r+

bors commented:
:pushpin: Commit fc8e10d5965c1d4954706cc080ba3a1b0bf06879 has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis fc8e10d5965c1d4954706cc080ba3a1b0bf06879 -->
<!-- homu: {"type":"Approved","sha":"fc8e10d5965c1d4954706cc080ba3a1b0bf06879","approver":"nikomatsakis"} -->

bors commented:
:umbrella: The latest upstream changes (presumably #72927) made this pull request unmergeable. Please resolve the merge conflicts.

jackh726 commented:
I rebased. Can I get this r+ed again? @nikomatsakis

Aaron1011 commented:
@jackh726: It looks like you've modified the `llvm-project` submodule: was that intentional?

jackh726 commented:
@Aaron1011 😔 no that happened during rebase. Fixed.

nikomatsakis commented:
@bors r+

bors commented:
:pushpin: Commit 4cf2833cf4a7fe7e283e9fc4d96ad124705a678b has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis 4cf2833cf4a7fe7e283e9fc4d96ad124705a678b -->
<!-- homu: {"type":"Approved","sha":"4cf2833cf4a7fe7e283e9fc4d96ad124705a678b","approver":"nikomatsakis"} -->

Dylan-DPC commented:
@bors p=1

bors commented:
:hourglass: Testing commit 4cf2833cf4a7fe7e283e9fc4d96ad124705a678b with merge d73e24b4f387054673f9ca9fa3a9deb3d6b73ba4...
<!-- homu: {"type":"BuildStarted","head_sha":"4cf2833cf4a7fe7e283e9fc4d96ad124705a678b","merge_sha":"d73e24b4f387054673f9ca9fa3a9deb3d6b73ba4"} -->

bors commented:
:broken_heart: Test failed - [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=31796)
<!-- homu: {"type":"BuildFailed","builder_url":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=31796","builder_name":"checks-azure"} -->

jackh726 commented:
Was this failure spurious?

Dylan-DPC commented:
@bors retry

bors commented:
:hourglass: Testing commit 4cf2833cf4a7fe7e283e9fc4d96ad124705a678b with merge 3a579a604437882c1af65f3e10212209831b9e67...
<!-- homu: {"type":"BuildStarted","head_sha":"4cf2833cf4a7fe7e283e9fc4d96ad124705a678b","merge_sha":"3a579a604437882c1af65f3e10212209831b9e67"} -->

bors commented:
:broken_heart: Test failed - [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=31820)
<!-- homu: {"type":"BuildFailed","builder_url":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=31820","builder_name":"checks-azure"} -->

jackh726 commented:
What the heck. Spurious? Network error?

Aaron1011 commented:
@bors retry

bors commented:
:hourglass: Testing commit 4cf2833cf4a7fe7e283e9fc4d96ad124705a678b with merge 4e345ba875018e384dd8ebd67a15ca1879cb0e1d...
<!-- homu: {"type":"BuildStarted","head_sha":"4cf2833cf4a7fe7e283e9fc4d96ad124705a678b","merge_sha":"4e345ba875018e384dd8ebd67a15ca1879cb0e1d"} -->

bors commented:
:broken_heart: Test failed - [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=31833)
<!-- homu: {"type":"BuildFailed","builder_url":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=31833","builder_name":"checks-azure"} -->

Aaron1011 commented:
For when the tree re-opens: @bors retry

bors commented:
:hourglass: Testing commit 4cf2833cf4a7fe7e283e9fc4d96ad124705a678b with merge 120287835df688101deea61657d1da5505a2e444...
<!-- homu: {"type":"BuildStarted","head_sha":"4cf2833cf4a7fe7e283e9fc4d96ad124705a678b","merge_sha":"120287835df688101deea61657d1da5505a2e444"} -->

Dylan-DPC commented:
yielding to imp pr

@bors retry

bors commented:
:hourglass: Testing commit 4cf2833cf4a7fe7e283e9fc4d96ad124705a678b with merge db94b8f7a92c1537d87e3a2e6694f1b57735d620...
<!-- homu: {"type":"BuildStarted","head_sha":"4cf2833cf4a7fe7e283e9fc4d96ad124705a678b","merge_sha":"db94b8f7a92c1537d87e3a2e6694f1b57735d620"} -->

RalfJung commented:
Yielding to rollup including this
@bors retry

bors commented:
:hourglass: Testing commit 4cf2833cf4a7fe7e283e9fc4d96ad124705a678b with merge 72b66d7efb19bc12c46b7944813785686688196c...
<!-- homu: {"type":"BuildStarted","head_sha":"4cf2833cf4a7fe7e283e9fc4d96ad124705a678b","merge_sha":"72b66d7efb19bc12c46b7944813785686688196c"} -->

RalfJung commented:
Actually, let's run this alone to see if it caused the OOM.

jackh726 commented:
@ralfjung that OOM happened previously on this PR. I thought it was spurious but maybe it isn't.

jackh726 commented:
Especially weird since I don't change any doc comments here. So I'm wondering if this related to the issue that I had seen before with the first Chalk integration PR, where there was an OOM using `parallel-compiler=true` when adding Chalk types to the arenas.

Dylan-DPC commented:
I'd prefer merging this separately. The queue is long at the moment and plus there are many other PRs that need to be merged individually so i'm gonna to delay this for a few days. If it is urgent, let us know

Dylan-DPC commented:
edit: just realised it is currently being merged, but in case it fails it might be on the queue for the few days

jackh726 commented:
This is fairly low priority. If merging fails, I think we can r- and I'll discuss with @nikomatsakis next week what we need to do going forward.

bors commented:
:broken_heart: Test failed - [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=31893)
<!-- homu: {"type":"BuildFailed","builder_url":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=31893","builder_name":"checks-azure"} -->

RalfJung commented:
Yeah same OOM again.
@bors rollup=never

Dylan-DPC commented:
created [zulip thread](https://rust-lang.zulipchat.com/#narrow/stream/144729-wg-traits/topic/Chalk.20upgrade.20failing.20on.20rustc.20due.20to.20oom) to discuss this further

Aaron1011 commented:
The OOM error is being caused by this loop never existing:

https://github.com/rust-lang/rust/blob/06e47688bf15d0215edbe05b21603062f6d2eb5d/src/librustc_trait_selection/traits/auto_trait.rs#L290

which causes us to keep pushing predicates into a `HashSet`. I'll work on a fix.

jackh726 commented:
Wow! Thanks  @Aaron1011

Aaron1011 commented:
The underlying cause of this infinite loop appears to be a bug in how `librustc_trait_selection` handles auto traits. On this branch, add the following code to `src/librustc_middle/traits/chalk.rs`

```rust
fn is_send<T: Send>() {}

fn bar<'tcx>() {
    is_send::<chalk_ir::VariableKind<RustInterner<'tcx>>>();
}
```

This will cause `rustc` to become stuck (with its memory usage growing forever) when building the **stage0** compiler. That is, this is *not* a bug in Chalk - it is a bug in the existing bootstrap compiler that (by an amazing coincidence) is getting triggered by some Chalk-related code that happens to have the right shape.

I haven't managed to create a minimal reproduction for this. However, I *did* find a minimized version of the `rustdoc` issue:

```rust
pub trait Interner {
    type InternedType;
}

struct RustInterner<'tcx> {
    foo: &'tcx ()
}

impl<'tcx> Interner for RustInterner<'tcx> {
    type InternedType = Box<TyData<Self>>;
}

enum TyData<I: Interner> {
    FnDef(I::InternedType)
}

struct VariableKind<I: Interner>(I::InternedType);

struct BoundVarsCollector<'tcx> {
    val: VariableKind<RustInterner<'tcx>>
}

fn is_send<T: Send>() {}

#[derive(Debug, Copy, Eq, Ord, Hash, PartialEq, PartialOrd, Clone)]
struct MyInterner<'tcx> {
    val: &'tcx ()
}

fn main() {
    is_send::<BoundVarsCollector<'static>>();
}
```

This code causes the infinite hang + memory growth on `rustdoc`. It hits the recursion limit with `rustc`, which is what *should* be happening with the code in `librustc_middle` (but isn't).

I captured a stacktrace from the running `rustc` process during compiler bootstrap:

<details>

```

#0  0x00007f158b59b2ed in rustc_data_structures::obligation_forest::ObligationForest<O>::register_obligation_at () from /home/aaron/repos/rust/build/x86_64-unknown-linux-gnu/stage0/bin/../lib/librustc_driver-3ec13ed752f7a54a.so
(gdb) bt
#0  0x00007f158b59b2ed in rustc_data_structures::obligation_forest::ObligationForest<O>::register_obligation_at () from /home/aaron/repos/rust/build/x86_64-unknown-linux-gnu/stage0/bin/../lib/librustc_driver-3ec13ed752f7a54a.so
#1  0x00007f158b59a461 in rustc_data_structures::obligation_forest::ObligationForest<O>::process_obligations () from /home/aaron/repos/rust/build/x86_64-unknown-linux-gnu/stage0/bin/../lib/librustc_driver-3ec13ed752f7a54a.so
#2  0x00007f158b4f6321 in <rustc_trait_selection::traits::fulfill::FulfillmentContext as rustc_infer::traits::engine::TraitEngine>::select_where_possible ()
   from /home/aaron/repos/rust/build/x86_64-unknown-linux-gnu/stage0/bin/../lib/librustc_driver-3ec13ed752f7a54a.so
#3  0x00007f1589e5d761 in rustc_typeck::check::FnCtxt::check_argument_types () from /home/aaron/repos/rust/build/x86_64-unknown-linux-gnu/stage0/bin/../lib/librustc_driver-3ec13ed752f7a54a.so
#4  0x00007f1589e0f324 in rustc_typeck::check::callee::<impl rustc_typeck::check::FnCtxt>::confirm_builtin_call () from /home/aaron/repos/rust/build/x86_64-unknown-linux-gnu/stage0/bin/../lib/librustc_driver-3ec13ed752f7a54a.so
#5  0x00007f1589e0d4f6 in rustc_typeck::check::callee::<impl rustc_typeck::check::FnCtxt>::check_call () from /home/aaron/repos/rust/build/x86_64-unknown-linux-gnu/stage0/bin/../lib/librustc_driver-3ec13ed752f7a54a.so
#6  0x00007f1589e1e81e in rustc_typeck::check::expr::<impl rustc_typeck::check::FnCtxt>::check_expr_kind () from /home/aaron/repos/rust/build/x86_64-unknown-linux-gnu/stage0/bin/../lib/librustc_driver-3ec13ed752f7a54a.so
#7  0x00007f1589e1de06 in rustc_typeck::check::expr::<impl rustc_typeck::check::FnCtxt>::check_expr_with_expectation_and_needs ()
   from /home/aaron/repos/rust/build/x86_64-unknown-linux-gnu/stage0/bin/../lib/librustc_driver-3ec13ed752f7a54a.so
#8  0x00007f1589e6151c in rustc_typeck::check::FnCtxt::check_stmt () from /home/aaron/repos/rust/build/x86_64-unknown-linux-gnu/stage0/bin/../lib/librustc_driver-3ec13ed752f7a54a.so
#9  0x00007f1589e61b5b in rustc_typeck::check::FnCtxt::check_block_with_expected () from /home/aaron/repos/rust/build/x86_64-unknown-linux-gnu/stage0/bin/../lib/librustc_driver-3ec13ed752f7a54a.so
#10 0x00007f1589e1e83a in rustc_typeck::check::expr::<impl rustc_typeck::check::FnCtxt>::check_expr_kind () from /home/aaron/repos/rust/build/x86_64-unknown-linux-gnu/stage0/bin/../lib/librustc_driver-3ec13ed752f7a54a.so
#11 0x00007f1589e1de06 in rustc_typeck::check::expr::<impl rustc_typeck::check::FnCtxt>::check_expr_with_expectation_and_needs ()
   from /home/aaron/repos/rust/build/x86_64-unknown-linux-gnu/stage0/bin/../lib/librustc_driver-3ec13ed752f7a54a.so
#12 0x00007f1589e28e6a in rustc_typeck::check::expr::<impl rustc_typeck::check::FnCtxt>::check_return_expr () from /home/aaron/repos/rust/build/x86_64-unknown-linux-gnu/stage0/bin/../lib/librustc_driver-3ec13ed752f7a54a.so
#13 0x00007f1589e4b29a in rustc_typeck::check::check_fn () from /home/aaron/repos/rust/build/x86_64-unknown-linux-gnu/stage0/bin/../lib/librustc_driver-3ec13ed752f7a54a.so
#14 0x00007f1589fe8346 in rustc_middle::ty::context::GlobalCtxt::enter_local () from /home/aaron/repos/rust/build/x86_64-unknown-linux-gnu/stage0/bin/../lib/librustc_driver-3ec13ed752f7a54a.so
#15 0x00007f1589e495f1 in rustc_typeck::check::typeck_tables_of () from /home/aaron/repos/rust/build/x86_64-unknown-linux-gnu/stage0/bin/../lib/librustc_driver-3ec13ed752f7a54a.so
#16 0x00007f1589fe585e in rustc_middle::ty::query::<impl rustc_query_system::query::config::QueryAccessors<rustc_middle::ty::context::TyCtxt> for rustc_middle::ty::query::queries::typeck_tables_of>::compute ()
   from /home/aaron/repos/rust/build/x86_64-unknown-linux-gnu/stage0/bin/../lib/librustc_driver-3ec13ed752f7a54a.so
#17 0x00007f1589ef6d68 in rustc_middle::dep_graph::<impl rustc_query_system::dep_graph::DepKind for rustc_middle::dep_graph::dep_node::DepKind>::with_deps ()
   from /home/aaron/repos/rust/build/x86_64-unknown-linux-gnu/stage0/bin/../lib/librustc_driver-3ec13ed752f7a54a.so
#18 0x00007f158a006887 in rustc_query_system::dep_graph::graph::DepGraph<K>::with_task_impl () from /home/aaron/repos/rust/build/x86_64-unknown-linux-gnu/stage0/bin/../lib/librustc_driver-3ec13ed752f7a54a.so
#19 0x00007f1589f278ea in rustc_data_structures::stack::ensure_sufficient_stack () from /home/aaron/repos/rust/build/x86_64-unknown-linux-gnu/stage0/bin/../lib/librustc_driver-3ec13ed752f7a54a.so
#20 0x00007f1589fb69c0 in rustc_query_system::query::plumbing::get_query_impl () from /home/aaron/repos/rust/build/x86_64-unknown-linux-gnu/stage0/bin/../lib/librustc_driver-3ec13ed752f7a54a.so
#21 0x00007f1589fbfabd in rustc_query_system::query::plumbing::ensure_query_impl () from /home/aaron/repos/rust/build/x86_64-unknown-linux-gnu/stage0/bin/../lib/librustc_driver-3ec13ed752f7a54a.so
#22 0x00007f1589e48e56 in rustc_typeck::check::typeck_item_bodies () from /home/aaron/repos/rust/build/x86_64-unknown-linux-gnu/stage0/bin/../lib/librustc_driver-3ec13ed752f7a54a.so
#23 0x00007f158a07a93e in rustc_middle::ty::query::<impl rustc_query_system::query::config::QueryAccessors<rustc_middle::ty::context::TyCtxt> for rustc_middle::ty::query::queries::typeck_item_bodies>::compute ()
   from /home/aaron/repos/rust/build/x86_64-unknown-linux-gnu/stage0/bin/../lib/librustc_driver-3ec13ed752f7a54a.so
#24 0x00007f1589ef6696 in rustc_middle::dep_graph::<impl rustc_query_system::dep_graph::DepKind for rustc_middle::dep_graph::dep_node::DepKind>::with_deps ()
   from /home/aaron/repos/rust/build/x86_64-unknown-linux-gnu/stage0/bin/../lib/librustc_driver-3ec13ed752f7a54a.so
#25 0x00007f158a00b394 in rustc_query_system::dep_graph::graph::DepGraph<K>::with_task_impl () from /home/aaron/repos/rust/build/x86_64-unknown-linux-gnu/stage0/bin/../lib/librustc_driver-3ec13ed752f7a54a.so
#26 0x00007f1589f18002 in rustc_data_structures::stack::ensure_sufficient_stack () from /home/aaron/repos/rust/build/x86_64-unknown-linux-gnu/stage0/bin/../lib/librustc_driver-3ec13ed752f7a54a.so
#27 0x00007f1589f8ff05 in rustc_query_system::query::plumbing::get_query_impl () from /home/aaron/repos/rust/build/x86_64-unknown-linux-gnu/stage0/bin/../lib/librustc_driver-3ec13ed752f7a54a.so
#28 0x00007f158a0618b8 in rustc_typeck::check_crate () from /home/aaron/repos/rust/build/x86_64-unknown-linux-gnu/stage0/bin/../lib/librustc_driver-3ec13ed752f7a54a.so
#29 0x00007f15896417b2 in rustc_interface::passes::analysis () from /home/aaron/repos/rust/build/x86_64-unknown-linux-gnu/stage0/bin/../lib/librustc_driver-3ec13ed752f7a54a.so
#30 0x00007f158946bedb in rustc_middle::ty::query::<impl rustc_query_system::query::config::QueryAccessors<rustc_middle::ty::context::TyCtxt> for rustc_middle::ty::query::queries::analysis>::compute ()
   from /home/aaron/repos/rust/build/x86_64-unknown-linux-gnu/stage0/bin/../lib/librustc_driver-3ec13ed752f7a54a.so
#31 0x00007f158945bbf8 in rustc_middle::dep_graph::<impl rustc_query_system::dep_graph::DepKind for rustc_middle::dep_graph::dep_node::DepKind>::with_deps ()
   from /home/aaron/repos/rust/build/x86_64-unknown-linux-gnu/stage0/bin/../lib/librustc_driver-3ec13ed752f7a54a.so
#32 0x00007f15894f0c04 in rustc_query_system::dep_graph::graph::DepGraph<K>::with_task_impl () from /home/aaron/repos/rust/build/x86_64-unknown-linux-gnu/stage0/bin/../lib/librustc_driver-3ec13ed752f7a54a.so
#33 0x00007f15894fe92e in rustc_query_system::query::plumbing::get_query_impl () from /home/aaron/repos/rust/build/x86_64-unknown-linux-gnu/stage0/bin/../lib/librustc_driver-3ec13ed752f7a54a.so
#34 0x00007f158946c6e7 in rustc_middle::ty::context::tls::enter_global () from /home/aaron/repos/rust/build/x86_64-unknown-linux-gnu/stage0/bin/../lib/librustc_driver-3ec13ed752f7a54a.so
#35 0x00007f158946e3d8 in rustc_interface::interface::run_compiler_in_existing_thread_pool () from /home/aaron/repos/rust/build/x86_64-unknown-linux-gnu/stage0/bin/../lib/librustc_driver-3ec13ed752f7a54a.so
#36 0x00007f15894ab02f in rustc_ast::attr::with_globals () from /home/aaron/repos/rust/build/x86_64-unknown-linux-gnu/stage0/bin/../lib/librustc_driver-3ec13ed752f7a54a.so
#37 0x00007f15894ad01e in std::sys_common::backtrace::__rust_begin_short_backtrace () from /home/aaron/repos/rust/build/x86_64-unknown-linux-gnu/stage0/bin/../lib/librustc_driver-3ec13ed752f7a54a.so
#38 0x00007f158945de6e in core::ops::function::FnOnce::call_once{{vtable-shim}} () from /home/aaron/repos/rust/build/x86_64-unknown-linux-gnu/stage0/bin/../lib/librustc_driver-3ec13ed752f7a54a.so
#39 0x00007f1588d7f81a in <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once () at /rustc/b7dc83a3f6ca9746fb3d121761c3605477b77d90/src/liballoc/boxed.rs:1076
#40 <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once () at /rustc/b7dc83a3f6ca9746fb3d121761c3605477b77d90/src/liballoc/boxed.rs:1076
#41 std::sys::unix::thread::Thread::new::thread_start () at src/libstd/sys/unix/thread.rs:87
#42 0x00007f1588c83422 in start_thread () from /usr/lib/libpthread.so.0
#43 0x00007f1588ba1bf3 in clone () from /usr/lib/libc.so.6

```
</details>

Aaron1011 commented:
I wonder if we could avoid this kind of bug with `rustc_trait_selection` by creating a new implementation of the trait system. Perhaps it could live in a separate crate. Maybe it could convert a trait predicates into some kind of logical inference rule...

Aaron1011 commented:
This is not *actually* a bug in `rustc` - it's just that the memory usage required by the trait solver grows *much* faster than the recorded depth (which is checked against `recursion_limit`). With `#![recursion_limit = "80"]` in `librustc_middle`, memory usage grew to 5GB before an overflow error was reported.

`librustc_middle` currently has a recursion limit of 512, so all of my computer's memory would likely be consumed before `rustc` ever managed to report an overflow error.

The `rustdoc` hang is definitely a bug - I'm working on a fix for it.

jackh726 commented:
I'm somewhat curious why there isn't a problem on master, but somewhere in the diff of this PR, the  OOM happens. I don't see any particular pattern in your repro @Aaron1011 that *isn't* present on current master.

jackh726 commented:
Blocked on #73452

jackh726 commented:
@rustbot modify labels to -S-waiting-on-review, +S-blocked

matthewjasper commented:
@bors r- (bors got restarted)

bors commented:
:umbrella: The latest upstream changes (presumably #73486) made this pull request unmergeable. Please resolve the merge conflicts.

Aaron1011 commented:
My minimized example from https://github.com/rust-lang/rust/pull/73335 no longer causes Rustdoc to OOM after https://github.com/rust-lang/rust/pull/73452 was merged. This PR should now be unblocked.

jackh726 commented:
@nikomatsakis care to r+ again?

Dylan-DPC commented:
@bors r=nikomatsakis

bors commented:
:pushpin: Commit d63195b0bc581837f77a22f46f2aa2303b911e56 has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis d63195b0bc581837f77a22f46f2aa2303b911e56 -->
<!-- homu: {"type":"Approved","sha":"d63195b0bc581837f77a22f46f2aa2303b911e56","approver":"nikomatsakis"} -->

bors commented:
:hourglass: Testing commit d63195b0bc581837f77a22f46f2aa2303b911e56 with merge a8cf3991177f30694200002cd9479ffbbe6d9a1a...
<!-- homu: {"type":"BuildStarted","head_sha":"d63195b0bc581837f77a22f46f2aa2303b911e56","merge_sha":"a8cf3991177f30694200002cd9479ffbbe6d9a1a"} -->

bors commented:
:sunny: Test successful - [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=32578)
Approved by: nikomatsakis
Pushing a8cf3991177f30694200002cd9479ffbbe6d9a1a to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"nikomatsakis","base_ref":"master","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=32578"},"merge_sha":"a8cf3991177f30694200002cd9479ffbbe6d9a1a"} -->

nnethercote commented:
This was a moderate [perf win](https://perf.rust-lang.org/compare.html?start=349f6bfb11d73ebb6a272f9a3d00883484f8218c&end=a8cf3991177f30694200002cd9479ffbbe6d9a1a&stat=instructions:u). Nice job!

jackh726 commented:
@nnethercote that's really unexpected, since basically all of this code doesn't run unless you're in `-Z chalk`. The only thing I can imagine giving a perf win would be the `rustc_hash` bump.
